common/io.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/io.hpp>
+
+namespace mender::common::io {
+
+StdError Copy(Writer &dst, Reader &src) {
+	vector<uint8_t> buffer(4096);
+	return Copy(dst, src, buffer);
+}
+
+StdError Copy(Writer &dst, Reader &src, vector<uint8_t> &buffer) {
+	size_t orig_size = buffer.size();
+
+	while (true) {
+		if (buffer.size() != orig_size) {
+			buffer.resize(orig_size);
+		}
+
+		auto result = src.Read(buffer);
+		if (!result) {
+			return result.error();
+		} else if (result.value() == 0) {
+			return StdError();
```

Comment:
Which `error_condition` is this and why no message?

Reply:
This again relates to the discussion about success value: It makes a zero error code object which means no error. `std` does not provide a macro for this AFAICS, but they do document that 0 means success (indirectly: `error_condition` comes from `std::errc` which has the semantics of `errno` and can be compared to it).

Suggestion: Maybe define zero as the success code ourselves, and then use that in construction?

Reply:
I like that suggestion better :)